dags/common/enricher.py

```diff
@@ -0,0 +1,106 @@
+            )
+
+        def clean_arxiv(arxiv):
+            if arxiv is None:
+                return None
+
+            return arxiv.split(":")[-1].split("v")[0].split(" ")[0].strip("\"'")
+
+        def get_arxiv_categories_from_response_xml(xml: ET.Element):
+            xml_namespaces = {
+                "arxiv": "http://arxiv.org/schemas/atom",
+                "w3": "http://www.w3.org/2005/Atom",
+            }
+
+            entries = xml.findall("./w3:entry", namespaces=xml_namespaces)
+            if len(entries) != 1:
+                return []
+            entry = entries[0]
+
+            primary_categories = [
+                node.attrib["term"]
+                for node in entry.findall(
+                    "./arxiv:primary_category", namespaces=xml_namespaces
+                )
+            ]
+            if not primary_categories:
+                return []
+            if len(primary_categories) > 1:
+                self.logger.error(
+                    f"Arxiv returned {len(primary_categories)} primary categories."
+                )
+                return []
+            primary_category = primary_categories[0]
+
+            secondary_categories = [
+                node.attrib["term"]
+                for node in entry.findall("./w3:category", namespaces=xml_namespaces)
+            ]
+
+            return list(set([primary_category] + secondary_categories))
+
+        arxiv_id = clean_arxiv(arxiv_id)
+
+        search_query = []
+
+        if title:
+            title = title.replace("-", "?").encode("ascii", "replace").decode()
+            search_query.append(f'ti:"{title}"')
+
+        params = {}
+        if arxiv_id:
+            params["id_list"] = arxiv_id
+        if search_query:
+            params["search_query"] = "".join(search_query)
+
+        request = requests.PreparedRequest()
+        request.prepare_url("http://export.arxiv.org/api/query", params)
+        response = requests.get(request.url)
+
+        categories = []
+        if response.status_code == 200:
+            xml = ET.fromstring(response.content)
+            categories = get_arxiv_categories_from_response_xml(xml)
+            if not categories:
+                self.logger.warning(
+                    f"Could not get arxiv categories for id={arxiv_id} title={title} doi={doi}"
+                )
+        else:
+            self.logger.error(
+                f"Got status_code {response.status_code} from arXiv when looking for categires for id={arxiv_id} title={title} doi={doi}"
+            )
+            response.raise_for_status()
+        return categories
+
+    def __call__(self, article: Dict):
+        enriched_article = article.copy()
+        enriched_article.update(
+            {
+                "$schema": self._get_schema(),
+                "categories": self._get_arxiv_categories(
+                    article.get("external_id", None)
+                ),
```

Comment:
shouldn't update directly the `arxiv_eprints` field?

Reply:
Yes good point. Changed 👍